File: app/src/main/java/com/mandarin/bcu/androidutil/stage/adapters/StageListAdapter.java
package com.mandarin.bcu.androidutil.stage.adapters;

import android.app.Activity;
import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import com.google.android.flexbox.FlexboxLayout;
import com.mandarin.bcu.R;
import com.mandarin.bcu.androidutil.StaticStore;

import java.util.ArrayList;
import java.util.List;

import common.util.stage.MapColc;
import common.util.stage.SCDef;
import common.util.stage.Stage;
import common.util.stage.StageMap;

public class StageListAdapter extends ArrayAdapter<String> {
    private final Activity activity;
    private final String[] stages;

    private final int mapcode;
    private final int stid;

    public StageListAdapter(Activity activity, String [] stages, int mapcode, int stid) {
        super(activity, R.layout.stage_list_layout,stages);

        this.activity = activity;
        this.stages = stages;
        this.mapcode = mapcode;
        this.stid = stid;
    }

    private static class ViewHolder {
        TextView name;
        FlexboxLayout icons;
        List<ImageView> images = new ArrayList<>();

        private ViewHolder(View row) {
            name = row.findViewById(R.id.stagename);
            icons = row.findViewById(R.id.enemicon);
        }
    }

    @NonNull
    public View getView(int position, View view, @NonNull ViewGroup parent) {
        View row;
        ViewHolder holder;

        if(view == null) {
            LayoutInflater inflater = LayoutInflater.from(activity);
            row = inflater.inflate(R.layout.stage_list_layout,parent,false);
            holder = new ViewHolder(row);
            row.setTag(holder);
        } else {
            row = view;
            holder = (ViewHolder)row.getTag();
        }

        holder.name.setText(stages[position]);

        holder.images.clear();
        holder.icons.removeAllViews();

        MapColc mc = StaticStore.map.get(mapcode);

        if(mc == null) return row;

        StageMap stm = mc.maps[stid];

        Stage st = stm.list.get(position);

        List<Integer> ids = getid(st.data);

        ImageView[] icons = new ImageView[ids.size()];

        for (int i = 0; i < ids.size(); i++) {
            icons[i] = new ImageView(activity);
            icons[i].setLayoutParams(new FlexboxLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT));
            icons[i].setImageBitmap(StaticStore.eicons[ids.get(i)]);
            icons[i].setPadding(StaticStore.dptopx(12f,activity),StaticStore.dptopx(4f,activity),0,StaticStore.dptopx(4f,activity));


            holder.icons.addView(icons[i]);
            holder.images.add(icons[i]);
        }

        return row;
    }

    private List<Integer> getid(SCDef stage) {
        List<int []> result = new ArrayList<>();
        int [][] data = reverse(stage.datas);

        for (int[] datas : data) {
            if (result.isEmpty()) {
                result.add(datas);
                continue;
            }

            int id = datas[SCDef.E];

            if (haveSame(id, result)) {
                result.add(datas);
            }
        }

        List<Integer> ids = new ArrayList<>();

        for(int [] datas : result) {
            ids.add(datas[SCDef.E]);
        }

        return ids;
    }

    private boolean haveSame(int id, List<int []> result) {
        if(id == 19 || id == 20 || id == 21) return false;

        for(int [] data : result) {
            if(id == data[SCDef.E])
                return false;
        }

        return true;
    }

    private int[][] reverse(int [][] data) {
        int [][] result = new int[data.length][];

        for(int i = 0; i < data.length; i++) {
            result[i] = data[data.length-1-i];
        }

        return result;
    }
}

File: app/src/main/java/com/mandarin/bcu/androidutil/stage/adapters/StageListAdapter.kt
package com.mandarin.bcu.androidutil.stage.adapters

import android.app.Activity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.ImageView
import android.widget.TextView
import com.google.android.flexbox.FlexboxLayout
import com.mandarin.bcu.R
import com.mandarin.bcu.androidutil.StaticStore
import common.pack.Identifier
import common.util.lang.MultiLangCont
import common.util.stage.SCDef
import common.util.stage.Stage
import common.util.unit.AbEnemy

class StageListAdapter(private val activity: Activity, private val stages: Array<Identifier<Stage>>) : ArrayAdapter<Identifier<Stage>>(activity, R.layout.stage_list_layout, stages) {

    private class ViewHolder constructor(row: View) {
        var name: TextView = row.findViewById(R.id.stagename)
        var icons: FlexboxLayout = row.findViewById(R.id.enemicon)
        var images: MutableList<ImageView?> = ArrayList()

    }

    override fun getView(position: Int, view: View?, parent: ViewGroup): View {
        val holder: ViewHolder
        val row: View

        if(view == null) {
            val inf = LayoutInflater.from(context)
            row = inf.inflate(R.layout.stage_list_layout,parent,false)
            holder = ViewHolder(row)
            row.tag = holder
        } else {
            row = view
            holder = row.tag as ViewHolder
        }

        val st = Identifier.get(stages[position]) ?: return row

        println(MultiLangCont.get(st))

        holder.name.text = MultiLangCont.get(st) ?: st.name ?: getStageName(stages[position].id)
        holder.images.clear()
        holder.icons.removeAllViews()

        val ids = getid(st.data)

        val icons = arrayOfNulls<ImageView>(ids.size)

        for (i in ids.indices) {
            icons[i] = ImageView(activity)
            icons[i]?.layoutParams = FlexboxLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT)
            if(ids[i].id < StaticStore.eicons?.size ?: 0) {
                val ic = StaticStore.eicons?.get(ids[i].id) ?: StaticStore.empty(1,1)
                icons[i]?.setImageBitmap(ic)
            } else {
                icons[i]?.setImageBitmap(StaticStore.empty(context, 18f, 18f))
            }
            icons[i]?.setPadding(StaticStore.dptopx(12f, activity), StaticStore.dptopx(4f, activity), 0, StaticStore.dptopx(4f, activity))
            holder.icons.addView(icons[i])
            holder.images.add(icons[i])
        }

        return row
    }

    private fun getid(stage: SCDef): List<Identifier<AbEnemy>> {
        val result: MutableList<SCDef.Line?> = ArrayList()
        val data = reverse(stage.datas)
        for (datas in data) {
            if (result.isEmpty()) {
                result.add(datas)
                continue
            }
            val id = datas!!.enemy
            if (haveSame(id, result)) {
                result.add(datas)
            }
        }
        val ids: MutableList<Identifier<AbEnemy>> = ArrayList()
        for (datas in result) {
            datas ?: continue
            ids.add(datas.enemy)
        }
        return ids
    }

    private fun haveSame(id: Identifier<AbEnemy>, result: List<SCDef.Line?>): Boolean {
        if (id.pack == Identifier.DEF && (id.id == 19 || id.id == 20 || id.id == 21))
            return false

        for (data in result) {
            data ?: continue
            if (id.equals(data.enemy)) return false
        }
        return true
    }

    private fun reverse(data: Array<SCDef.Line>): Array<SCDef.Line?> {
        val result = arrayOfNulls<SCDef.Line>(data.size)
        for (i in data.indices) {
            result[i] = data[data.size - 1 - i]
        }
        return result
    }

    private fun getStageName(num: Int) : String {
        return "Stage"+number(num)
    }

    private fun number(num: Int): String {
        return if (num in 0..9) "00$num" else if (num in 10..99) "0$num" else "" + num
    }

}